### PR TITLE
Implement ArgMax operation

### PR DESCRIPTION
**Pull request by shizy818**

### CODE REVIEW

File: src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp
Comment (chentong319) [inline src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp:74]: It seems that you are using the old interface for krnl loops. The new interface will construct the Krnl loop nest as an object and operate on it, such as push bound and get induction variable. Which document did you use for Krnl dialect?

Comment (shizy818) [reply src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp:74]: I think you mean class `BuildKrnlLoop`. I don't know there is a Krnl document. I read code `ONNXFlattenOpLowering` and `ONNXReductionOpLowering` and do similar process for `ONNXArgMaxOpLowering`. BTW, it seems they are using the old interface.

I can change to use `BuildKrnlLoop`. If there is a guide document for krnl dialect, please let me know. Thanks!

Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp:74]: Yes, it is BuildKrnlLoop. 
We will discuss how to provide the document and get back to you.

**Comment by tungld** [comment]
@shizy818 could you please add lit tests for the onnx-to-krnl lowering in `test/mlir/onnx/onnx_lowering.mlir`? Thanks!

File: src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp:74]: 1. We think it is necessary to use the BuildKrnlLoop interface for the newly added code.
2. Sorry for that there is no proper document yet
3. You can follow the examples for other operations (such as Tile). Notice that the newly added code uses another utility, called IndexExpr. It is not absolutely needed in your case, but may be helpful in simplifying the code for mermen allocation/loop bound,  and the same logic can be used for shape inference too.
Any question, let me know.

**Comment by shizy818** [comment]
@chentong319 @tungld got it.

File: test/backend/test.py
Comment (chentong319) [inline test/backend/test.py:130]: You did add the backend test. It seems that the Krnl dialect fixed the issue I mentioned above.

File: src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp
Comment (shizy818) [reply src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp:28]: It seems I can remove L28 and all tests are OK. However I see it in many cases including `Tile` op. Do you know why some other ops need it?

File: src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp
Comment (shizy818) [reply src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp:81]: Change to use `saveInsertionPoint` & `restoreInsertionPoint`. Thanks!

File: test/backend/test.py
Comment (shizy818) [reply test/backend/test.py:130]: So I need not use affine expression instead?

File: src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp:28]: I found the comment for such code: "Add void cast to supress assert related unused warnings"

Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp:28]: Keep it.

File: src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp:81]: I meant to use saveInsertionPoint() before you change insertion point to the loop body. The intention is to avoid the assumption the insert point is the parent block of current insertion place.

File: test/backend/test.py
Comment (chentong319) [reply test/backend/test.py:130]: It seems that you do not need to.

**Comment by gongsu832** [comment]
test this please

**Comment by chentong319** [comment]
@gongsu832 Is the failure caused by out-of-date branch?

**Review by chentong319 — COMMENTED** [review]
Thanks for the changes.
Please merge to the master version. 
I will approve this patch after it passes the tests.

**Comment by gongsu832** [comment]
@chentong319 The patch is failing a lit test.

**Comment by chentong319** [comment]
@shizy818 Did check-onnx-lit fail in your local build? You can try "make check-onnx-lit", as well as "make check-onnx-backend" and "make check-onnx-backend-dynamic".

**Comment by shizy818** [comment]
> @shizy818 Did check-onnx-lit fail in your local build? You can try "make check-onnx-lit", as well as "make check-onnx-backend" and "make check-onnx-backend-dynamic".

It runs clean on my local env macOS 10.15.7
```
Testing Time: 1.65s
  Unsupported:  1
  Passed     : 38
[100%] Built target check-onnx-lit
```

I will rebase on latest master branch and see if it is OK. BTW, which test failed now? "check-onnx-lit" or "check-onnx-backend"?

File: test/mlir/onnx/onnx_lowering.mlir
Comment (tungld) [inline test/mlir/onnx/onnx_lowering.mlir:3473]: @shizy818 Please remove " from `slt` and `ogt` to avoid errors in `check-onnx-lit`, e.g. `cmpi "slt"` to `cmpi slt`.

Comment (tungld) [reply test/mlir/onnx/onnx_lowering.mlir:3473]: This change is because of our recent LLVM commit update.

Comment (shizy818) [reply test/mlir/onnx/onnx_lowering.mlir:3473]: I see. Thx

**Review by chentong319 — APPROVED** [review]
LGTM!

**Comment by chentong319** [comment]
@gongsu832 The test on linux-s390x did not come back for almost one day. Do you know why?

**Comment by gongsu832** [comment]
@chentong319 The linux-s390x test uses polling to detect if anything changed for a PR to trigger a (re)build. Jenkins log shows that it has stopped detecting changes for this particular PR, which I don't know why. Perhaps there is some bug in the GitHub Pull Request Builder plugin's algorithm for detecting changes.
```
2021-02-16 05:26:00.885+0000 [id=57]    INFO    o.j.p.ghprb.GhprbPullRequest#updatePR: Pull request #510 was updated on repo onnx/onnx-mlir but there aren't any new comments nor commits; that may mean that commit status was updated.
```
In any case, we will not be using this job anymore. It will be replaced by the new dockerized build job.

@tjingrant I think we can now switch over to the new dockerized build jobs, making them required. I will disable the linux-s390x job (or you can do that) once you deselect it from the GitHub required list.

**Comment by shizy818** [comment]
@gongsu832 It looks there is nothing I can do about this. Is it because I do too much force push?

**Comment by gongsu832** [comment]
@shizy818 It's not clear what exactly caused linux-s390x to stop detecting changes. But as I mentioned in my previous message, we will not be using linux-s390x anymore. So Tian should remove linux-s390x as required test and make the new dockerized Jenkins tests as required. Once that's done, we will be able to merge.

**Pull request merged**
